src/pymor/algorithms/rand_la.py

```diff
@@ -0,0 +1,362 @@
+    maxnorm = np.inf
+    M = A.apply(R)
+
+    while(maxnorm > testlimit):
+        basis_length = len(B)
+        v = A.source.random(distribution='normal')
+        if iscomplex:
+            v += 1j*A.source.random(distribution='normal')
+        B.append(A.apply(v))
+        gram_schmidt(B, range_product, atol=0, rtol=0, offset=basis_length, copy=False)
+        M -= B.lincomb(B.inner(M, range_product).T)
+        maxnorm = np.max(M.norm(range_product))
+
+    return B
+
+
+@defaults('q', 'l')
+def rrf(A, source_product=None, range_product=None, q=2, l=8, return_rand=False, iscomplex=False):
+    """Randomized range approximation of `A`.
+
+    This is an implementation of Algorithm 4.4 in :cite:`HMT11`.
+
+    Given the |Operator| `A`, the return value of this method is the |VectorArray|
+    `Q` whose vectors form an approximate orthonomal basis for the range of `A`.
+
+    Parameters
+    ----------
+    A
+        The |Operator| A.
+    source_product
+        Inner product |Operator| of the source of A.
+    range_product
+        Inner product |Operator| of the range of A.
+    q
+        The number of power iterations.
+    l
+        The block size of the normalized power iterations.
+    return_rand
+        If 'True', the randomly sampled |VectorArray| R is returned. 
+    iscomplex
+        If `True`, the random vectors are chosen complex.
+
+    Returns
+    -------
+    Q
+        |VectorArray| which contains the basis, whose span approximates the range of A.
+    R
+        The randomly sampled |VectorArray|
+    """
+    assert isinstance(A, Operator)
+    
+    if range_product is None:
+        range_product = IdentityOperator(A.range)
+    else:
+        assert isinstance(range_product, Operator)
+        
+    if source_product is None:
+        source_product = IdentityOperator(A.source)
+    else:
+        assert isinstance(source_product, Operator)
+    
+    R = A.source.random(l, distribution='normal')
+
+    if iscomplex:
+        R += 1j*A.source.random(l, distribution='normal')
+
+    Q = A.apply(R)
+    gram_schmidt(Q, range_product, atol=0, rtol=0, copy=False)
+
+    for i in range(q):
+        Q = A.apply_adjoint(range_product.apply(Q))
+        Q = source_product.apply_inverse(Q)
+        gram_schmidt(Q, source_product, atol=0, rtol=0, copy=False)
+        Q = A.apply(Q)
+        gram_schmidt(Q, range_product, atol=0, rtol=0, copy=False)
+
+    if return_rand:
+        return Q, R
+    else:
+        return Q
+
+
+@defaults('p', 'q', 'modes')
+def random_generalized_svd(A, range_product=None, source_product=None, modes=3, p=20, q=2):
+    """Randomized SVD of a |VectorArray| or an |Operator|. 
+
+    Saibaba AK, Hart J, van BloemenWaanders B. Randomized algorithms for
+    generalized singular value decomposition with application to sensitivity analysis. Numer Linear Algebra Appl.
+    2021;28:e2364. 
+    https://doi.org/10.1002/nla.2364
```

Comment:
Sorry if I was unclear earlier, I think it would be best to add an entry in `docs/source/bibliography.bib` and cite it here with `` :cite:`SHB21` ``  or similar

Reply:
This still needs to be fixed.